R/conditions.R

```diff
@@ -0,0 +1,815 @@
+#'   muffle a condition. It stops the condition from being passed to
+#'   other handlers when the inplace handler did not jump
+#'   elsewhere. \code{TRUE} by default for benign conditions, but
+#'   \code{FALSE} for critical ones, since in those cases execution
+#'   should probably not be allowed to continue normally.
+#' @seealso \code{\link{panic}()}, \code{\link{warn}()} and
+#'   \code{\link{inform}()} for signalling typical R conditions. See
+#'   \code{\link{with_handlers}()} for establishing condition
+#'   handlers.
+#' @export
+#' @examples
+#' # Creating a condition of type "foo"
+#' cnd <- cnd_new("foo")
+#'
+#' # If no handler capable of dealing with "foo" is established on the
+#' # stack, signalling the condition has no effect:
+#' cnd_signal(cnd)
+#'
+#' # To learn more about establishing condition handlers, see
+#' # documentation for with_handlers(), thrown() and inplace():
+#' with_handlers(cnd_signal(cnd),
+#'   foo = inplace(function(c) cat("side effect!\n"))
+#' )
+#'
+#'
+#' # By default, cnd_signal() creates a muffling restart which allows
+#' # inplace handlers to prevent a condition from being passed on to
+#' # other handlers and to resume execution:
+#' undesirable_handler <- inplace(function(c) cat("please don't call me\n"))
+#' muffling_handler <- inplace(function(c) {
+#'   cat("muffling foo...\n")
+#'   rst_muffle(c)
+#' })
+#'
+#' with_handlers(foo = undesirable_handler,
+#'   with_handlers(foo = muffling_handler, {
+#'     cnd_signal("foo")
+#'     "return value"
+#'   }))
+#'
+#'
+#' # You can signal a critical condition with cnd_panic(). Unlike
+#' # cnd_signal() which has no side effect besides signalling the
+#' # condition, cnd_panic() makes the program terminate with an error
+#' # unless a handler can deal with the condition:
+#' \dontrun{
+#' cnd_panic(cnd)
+#' }
+#'
+#' # If you don't specify a .msg or .call, the default message/call
+#' # (supplied to cnd_new()) are displayed. Otherwise, the ones
+#' # supplied to cnd_panic() and cnd_signal() take precedence:
+#' \dontrun{
+#' critical <- cnd_new("my_error",
+#'   .msg = "default 'my_error' msg",
+#'   .call = quote(default(call))
+#' )
+#' cnd_panic(critical)
+#' cnd_panic(critical, .msg = "overridden msg")
+#'
+#' fn <- function(...) {
+#'   cnd_panic(critical, .call = TRUE)
+#' }
+#' fn(arg = foo(bar))
+#' }
+#'
+#' # Note that by default a condition signalled with cnd_panic() does
+#' # not have a muffle restart. That is because in most cases,
+#' # execution should not continue after signalling a critical
+#' # condition.
+cnd_signal <- function(.cnd, ..., .msg = NULL, .call = FALSE,
+                       .mufflable = TRUE) {
+  call <- if (.call) sys.call(-1) else NULL
+  cnd <- make_cnd(.cnd, ..., .msg = .msg, .call = call)
+  cnd_signal_(cnd, base::signalCondition, .mufflable)
+}
+#' @rdname cnd_signal
+#' @export
+cnd_panic <- function(.cnd, ..., .msg = NULL, .call = FALSE,
```

Comment:
I was thinking creating a condition manually would be somewhat exceptional. So you'd use `cnd_signal()` and `cnd_panic()` to create a condition on the fly:

```r
# Create a condition of type `foo` and signal it:
cnd_signal("foo")
```

We could still switch within a single `cnd_signal()` function by accepting a chr vector:

```r
cnd_signal(c("my_type", "error"))
```

There is the issue that by default signalling a serious error should be done without a muffle restart. We could take `.mufflable = NULL`.